File: jdollarx/src/main/java/org/lab49/jdollarx/custommatchers/IsPresentNTimes.java
package org.lab49.jdollarx.custommatchers;

import org.lab49.jdollarx.InBrowser;
import org.lab49.jdollarx.Path;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import static org.lab49.jdollarx.custommatchers.CustomMatchersUtil.wrap;

public class IsPresentNTimes {
    private final int nTimes;

    public IsPresentNTimes(int nTimes){
        this.nTimes = nTimes;
    }

    public Matcher<Path> timesIn(InBrowser browser){
        return new TypeSafeMatcher<Path>() {
            private int foundNTimes;
            private Path el;

            @Override
            public void describeTo(final Description description) {
                 description.appendText( String.format("browser page contains %s %d time%s", wrap(el), nTimes, nTimes != 1 ? "s" : ""));
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(wrap(el) + " appears " + foundNTimes + " time" + (foundNTimes!=1 ? "s" : ""));
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                foundNTimes = browser.numberOfAppearances(el);
                return foundNTimes == nTimes;
            }
        };
    }
}

File: dollarx/src/main/java/com/github/loyada/dollarx/custommatchers/hamcrest/IsPresentNTimes.scala
package com.github.loyada.dollarx.custommatchers.hamcrest

import com.github.loyada.dollarx.custommatchers.CustomMatchersUtil
import com.github.loyada.dollarx.{PathParsers, RelationOperator, Path, Browser}
import org.hamcrest.{TypeSafeMatcher, Description, Matcher}
import org.openqa.selenium.NoSuchElementException
import org.w3c.dom.Document

case class IsPresentNTimes(nTimes: Int) {
  import RelationOperator._
  case class PresentNTimesMatcher(browser: Browser, relationOperator: RelationOperator = exactly) extends TypeSafeMatcher[Path]() {
    private var foundNTimes: Int = 0
    private var el: Path = null

    def describeTo(description: Description) {
      description.appendText(s"browser page contains ${CustomMatchersUtil.wrap(el)}${opAsEnglish(relationOperator)}$nTimes time${if (nTimes != 1) "s" else ""}")
    }

    protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
      mismatchDescription.appendText(CustomMatchersUtil.wrap(el) + " appears " + foundNTimes + " time" + (if (foundNTimes != 1) "s" else ""))
    }

    protected def matchesSafely(el: Path): Boolean = {
      this.el = el
      try {
        browser.findPageWithNumberOfOccurrences(el, nTimes, relationOperator)
        true
      }
      catch {
        case e: NoSuchElementException => {
          foundNTimes = browser.numberOfAppearances(el)
          false
        }
      }
    }
  }

  case class PresentNTimesMatcherForDocument(document: Document, relationOperator: RelationOperator = exactly) extends TypeSafeMatcher[Path]() {
    private var foundNTimes: Int = 0
    private var el: Path = null

    def describeTo(description: Description) {
      description.appendText(s"document contains ${CustomMatchersUtil.wrap(el)}${opAsEnglish(relationOperator)}$nTimes time${if (nTimes != 1) "s" else ""}")
    }

    protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
      mismatchDescription.appendText(CustomMatchersUtil.wrap(el) + " appears " + foundNTimes + " time" + (if (foundNTimes != 1) "s" else ""))
    }

    protected def matchesSafely(el: Path): Boolean = {
      this.el = el
      foundNTimes = PathParsers.findAllByPath(document, el).getLength
      relationOperator match {
          case RelationOperator.orLess =>  foundNTimes<=nTimes
          case RelationOperator.orMore => foundNTimes>=nTimes
          case RelationOperator.exactly => foundNTimes==nTimes
      }
    }
  }

  def timesIn(browser: Browser): Matcher[Path] = PresentNTimesMatcher(browser)
  def timesOrMoreIn(browser: Browser): Matcher[Path] = PresentNTimesMatcher(browser, orMore)
  def timesOrLessIn(browser: Browser): Matcher[Path] = PresentNTimesMatcher(browser, orLess)

  def timesIn(document: Document): Matcher[Path] = PresentNTimesMatcherForDocument(document)
  def timesOrMoreIn(document: Document): Matcher[Path] = PresentNTimesMatcherForDocument(document, orMore)
  def timesOrLessIn(document: Document): Matcher[Path] = PresentNTimesMatcherForDocument(document, orLess)

}
